cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/SingleThreadDatasetCache.java

```diff
@@ -194,7 +233,18 @@ public void dismissTransactionContext() {
 
   @Override
   public Iterable<TransactionAware> getTransactionAwares() {
-    return txContext == null ? NO_TX_AWARES : Iterables.concat(extraTxAwares, inProgress.values());
+    if (txContext == null) {
+      return NO_TX_AWARES;
+    }
+    ArrayList<TransactionAware> result = new ArrayList<>(extraTxAwares.size() + activeTxAwares.size());
+    result.addAll(extraTxAwares);
+    for (WeakReferenceDelegatingTransactionAware reference : activeTxAwares.values()) {
+      TransactionAware txAware = reference.get();
+      if (txAware != null) {
+        result.add(txAware);
```

Comment:
Yes, I can do that. This method is only used in tests, though. These tests bypass the tx-context of the cache and create their own tx-context. It should still work.